### PR TITLE
[docs]: Add specification and config docs

### LINKED ISSUE #275: [docs] Meta: Go through all types and configs and document
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**What happened**:
We currently lack a lot of documentation.

**What you expected to happen**:
Document each specific type and how to configure it from the operations side as well as from the TXT record side.
Additionally we need to document the additional operations features such as "qr", "metrics", "resolver" etc.

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Adds specification and config docs for all types.

TODO:

- [x] Gomods specification
- [x] Proxy description
- [x] Prometheus config
- [x] Gomods config
- [x] QR config
- [x] Resolver config
- [ ] wildcard documentation

**Which issue this PR fixes**:
fixes txtdirect/txtdirect#275

**Special notes for your reviewer**:
Will add all things that need documentation to the todo list to be sure we don't miss anything.

### CODE REVIEW

**Comment by erbesharat** [comment]
Docs are not ready yet and still need a bit of cleaning, but I wanted to be sure we're not missing anything so a quick review would be awesome.

I feel a few of the config sections still need some description but not sure which ones.

File: content/docs/configuration/content.md [hunk 1]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+  sticky = true
 +++
 
-**Redirect all requests to "www"-subdomain:**  
-*example.com -> www.example.com*
+# Redirect Types
+
+---
+
+## Supported redirect types:
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
## Supported redirect types
```

File: content/docs/configuration/content.md [hunk 2]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+
+| Type Name | Type Description                                            |
```
Comment (stp-ip) [inline content/docs/configuration/content.md:18]: This is mostly an issue with Syna not making markdown tables look pretty yet. You can bug @mpourismaiel about it ;)

File: content/docs/configuration/content.md [hunk 3]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+| --------- | ----------------------------------------------------------- |
+| www       | Redirect all requests to "www"-subdomain                    |
+| host      | Redirect to host provided in TXT record                     |
+| path      | Enable path based redirects                                 |
+| gometa    | Enable Go packages meta/vanity redirects                    |
```
Comment (stp-ip) [inline content/docs/configuration/content.md]: *Go package
 Singular feels better here.

File: content/docs/configuration/content.md [hunk 4]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+| gomods    | Enable serving and caching Go modules                       |
```
Comment (stp-ip) [inline content/docs/configuration/content.md:24]: We'll have to think about the inclusion of this within the docs update or, if it would be better to leave this out.

File: content/docs/configuration/content.md [hunk 5]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+| dockerv2  | Redirect to docker registry or image provided in TXT record |
+| proxy     | Proxy the request to endpoint provided in TXT record        |
+
+---
+
+## Enable type
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
## Enable single type
```

File: content/docs/configuration/content.md [hunk 6]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
 ```
 txtdirect {
-  enable host
+  enable host path gomods dockerv2
 }
 ```
 
-**Redirect to host provided in TXT record:**  
-*Fallback: Redirect to "www"-subdomain on empty record*
+## Disable type
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
## Disable specific type
```

File: content/docs/configuration/content.md [hunk 7]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
 
 [sidebar]
   enable = true
+  sticky = true
```
Comment (stp-ip) [inline content/docs/configuration/content.md:9]: Sticky navbar isn't scrollable. @mpourismaiel

Comment (stp-ip) [reply content/docs/configuration/content.md:9]: Also the indentation isn't working correctly. Levels are intended more than they should be.

File: content/docs/configuration/content.md [hunk 8]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+
+Step-by-Step fallback flow:
+
+- Specific record's `to=` field
+- Specific record's `website=` field
+- Specific record's `root=` field
+- Second to last record's `website=` field
+- Second to last record's `root=` field
+- Second to last record's `to=` field
+- Request's host "www"-subdomain
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
- Request's host "www"-subdomain (if enabled)
```

File: content/docs/configuration/content.md [hunk 9]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+- Global config's `redirect` field
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
- Global config's `redirect` field (if configured)
```

File: content/docs/configuration/content.md [hunk 10]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+- 404 Status code
+
+_hint: **Specific record** is the last TXT record found_
```
Comment (stp-ip) [inline content/docs/configuration/content.md]: Why use a hint? Instead of "specific record" we could say "last found TXT record" above. Maybe both are not clear, but a hint feels like a workaround, that doesn't resolve this.

Comment (stp-ip) [reply content/docs/configuration/content.md]: Ideas: (happy for other suggesitons)
`- Current TXT record *to=* field`
`- Parent (type=path) TXT record *to=* field`

File: content/docs/configuration/content.md [hunk 11]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+- Specific record's `root=` field
+- Second to last record's `website=` field
+- Second to last record's `root=` field
+- Second to last record's `to=` field
+- Request's host "www"-subdomain
+- Global config's `redirect` field
+- 404 Status code
+
+_hint: **Specific record** is the last TXT record found_  
+_hint 2: **Second to last record** is the last path record leading to the **Specific record**_
```
Comment (stp-ip) [inline content/docs/configuration/content.md]: Instead of a hint, maybe an example does explain it better?

File: content/docs/configuration/content.md [hunk 12]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+| ------------------------------------- | ------------------------------------------ |
+| txtdirect_redirect_count_total        | Total requests per host                    |
+| txtdirect_redirect_status_count_total | Total returned statuses per host           |
+| txtdirect_redirect_type_count_total   | Total requests for each host based on type |
+| txtdirect_fallback_type_count_total   | Total fallbacks triggered for each type    |
+| txtdirect_redirect_path_count_total   | Total redirects per path for each host     |
+
+## Prometheus config
+
+You can either use the default settings for enabling Promethues metrics or the advanced config to provide a custom address and path for exporting metrics.
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
You can either use the default settings for enabling Prometheus metrics or the advanced config to provide a custom address and path for exporting metrics.
```

File: content/docs/configuration/content.md [hunk 13]
```diff
@@ -6,58 +6,214 @@ title = "Configuration"
...
+    foreground #000000
+  }
+}
+```
+
+# DNS Resolver
+
+---
+
+To use a custom DNS resolver instead of machine's default resolver, you can provide the `resolver` address in the config.
```
Comment (stp-ip) [inline content/docs/configuration/content.md]:
```suggestion
To use a custom DNS resolver instead of the machine's default resolver, you can provide the `resolver` address in the config.
```

File: content/docs/specification/content.md [hunk 14]
```diff
@@ -143,7 +143,7 @@ If a custom regex is configured with the `re=` field each match is used as a sub
 
 - Key: **from**
 - Permitted values: "simplified regex"
-- Example: "from=/$2/$1/$3"
+- Example: "from=/$2/$1/\$3"
```
Comment (stp-ip) [inline content/docs/specification/content.md]: Why the change?

File: content/docs/specification/content.md [hunk 15]
```diff
@@ -276,23 +283,20 @@ Using `gometa` type you can point to a Go package inside your records using the
...
-- Mandatory
-- Permitted values: "absolute URL"
-- Example: "to=https://example.com/subpage/"
+# Proxy Type (experimental)
 
----
+## Description
 
-# Tor Type (experimental)
+The `proxy` type let's you to proxy your requests to a specific endpoint. It reads the endpoint from record's `to=` field and reverse proxies the requst to the endpoint.
```
Comment (stp-ip) [inline content/docs/specification/content.md]:
```suggestion
The `proxy` type let's you proxy your requests to a specific upstream. It reads the upstream from the record's `to=` field and reverse proxies the request.
```

I think upstream is more widely used. Another option would be backend.

**Review by stp-ip — CHANGES_REQUESTED** [review]
I would probably leave out the ToC as we do have the sticky nav on the left. Feels too much spaces is used for ToC before the user jumps into our actual docs. What do you think?

File: content/docs/specification/content.md (hunk 14 above)
Comment (erbesharat) [reply content/docs/specification/content.md]: Just like the last time in the previous PR. It's because of VSCode's Markdown extension.

File: content/docs/configuration/content.md (hunk 10 above)
Comment (erbesharat) [reply content/docs/configuration/content.md]: Since we only support `path` as a parent record for now, Think it would be fine to use the examples you mentioned.

File: content/docs/configuration/content.md (hunk 7 above)
Comment (erbesharat) [reply content/docs/configuration/content.md:9]: Added an issue for sticky bug in Syna.

**Comment by stp-ip** [comment]
Let's merge and do the wildcard docs in a follow on PR.

/lgtm

**Review by stp-ip — APPROVED** [review]
